clang/lib/Sema/SemaTemplateInstantiate.cpp

```diff
@@ -231,14 +231,18 @@ Response HandleFunctionTemplateDecl(const FunctionTemplateDecl *FTD,
                                     MultiLevelTemplateArgumentList &Result) {
   if (!isa<ClassTemplateSpecializationDecl>(FTD->getDeclContext())) {
     NestedNameSpecifier *NNS = FTD->getTemplatedDecl()->getQualifier();
-    const Type *Ty;
-    const TemplateSpecializationType *TSTy;
-    if (NNS && (Ty = NNS->getAsType()) &&
-        (TSTy = Ty->getAs<TemplateSpecializationType>()))
-      Result.addOuterTemplateArguments(const_cast<FunctionTemplateDecl *>(FTD),
-                                       TSTy->template_arguments(),
-                                       /*Final=*/false);
+
+    while (const Type *Ty = NNS ? NNS->getAsType() : nullptr) {
```

Comment:
This make sense, I was just looking at the AST: https://godbolt.org/z/P3brjcsMG

and I see that `InnerClass` has:

```console
| |-NestedNameSpecifier TypeSpec 'Base<T>'
```

and I wondering what the prefix is there? Maybe the AST is not telling the whole story?

Reply:
So that one you're seeing there is for the `InnerClass` on line 13, the `NestedNameSpecifier` is for 'everything else' outside of that declaration.  So in THAT case, the Prefix of that one is nothing, since the `Base<T>`  is the entirety of the thing.

However, that AST doesn't show the `NNS` of `func` on line 22, for some reason we're not dumping that.  In THAT case, the `NNS` is `Base<T>::InnerClass`.  At that point, the `InnerClass` doesn't have template arguments (since it isn't a template!), but its `Prefix` is just the `Base<T>`, which has them (which is why/how this loop works).

Reply:
It does not look like there is a simple way of getting like w/ `CXXRecordDecl` so some work would be involved.

Reply:
I don't understand what you mean?  `TagDecl` (IIRC?) has a `getQualifier()` to get the appropriate `NestedNameSpecifier`.  I believe our NNS tree is correct the way I've implemented it above, with the exception of the issue I posted about a moment ago.

Reply:
Yeah, I just realized I was looking at the code a moment ago `FTD->getTemplatedDecl()->getQualifier()`